### PR TITLE
[generator] Fixed foot/bicycle tag from Relation.

### PR DESCRIPTION
**Pull request by vng**
https://github.com/organicmaps/organicmaps/issues/8028

### CODE REVIEW

**Review by pastk — APPROVED** [review]

File: generator/osm_element.cpp [hunk 1]
```diff
@@ -71,22 +69,15 @@ void OsmElement::AddTag(char const * key, char const * value)
...
-}
-
-void OsmElement::AddTag(std::string const & key, std::string const & value)
-{
-  AddTag(key.data(), value.data());
+  m_tags.emplace_back(key, value);
 }
 
-bool OsmElement::HasTag(std::string const & key) const
+bool OsmElement::HasTag(std::string_view const & key) const
```
Comment (biodranik) [inline generator/osm_element.cpp:75]:
```suggestion
bool OsmElement::HasTag(std::string_view key) const
```

File: generator/osm_element.cpp [hunk 2]
```diff
@@ -71,22 +69,15 @@ void OsmElement::AddTag(char const * key, char const * value)
...
 {
   return base::AnyOf(m_tags, [&](auto const & t) { return t.m_key == key; });
 }
 
-bool OsmElement::HasTag(std::string const & key, std::string const & value) const
+bool OsmElement::HasTag(std::string_view const & key, std::string_view const & value) const
```
Comment (biodranik) [inline generator/osm_element.cpp:80]:
```suggestion
bool OsmElement::HasTag(std::string_view key, std::string_view value) const
```

File: generator/osm_element.hpp [hunk 3]
```diff
@@ -102,11 +102,10 @@ struct OsmElement
...
 
-  void AddTag(Tag const & tag);
-  void AddTag(char const * key, char const * value);
-  void AddTag(std::string const & key, std::string const & value);
-  bool HasTag(std::string const & key) const;
-  bool HasTag(std::string const & key, std::string const & value) const;
+  void AddTag(std::string_view key, std::string_view value);
+  void AddTag(Tag const & tag) { AddTag(tag.m_key, tag.m_value); }
+  bool HasTag(std::string_view const & key) const;
+  bool HasTag(std::string_view const & key, std::string_view const & value) const;
```
Comment (biodranik) [inline generator/osm_element.hpp:108]:
```suggestion
  bool HasTag(std::string_view key) const;
  bool HasTag(std::string_view key, std::string_view value) const;
```

**Review by biodranik — APPROVED** [review]

File: generator/osm_element.cpp (hunk 1 above)
Comment (vng) [reply generator/osm_element.cpp:75]: Like we always do ```m2::PointD const &```, I don't see a reason to bother/change here.
Anyway, both variants are ok.

File: generator/osm_element.cpp [hunk 4]
```diff
@@ -36,57 +35,43 @@ std::string DebugPrint(OsmElement::EntityType type)
...
-
-#define SKIP_KEY_BY_PREFIX(skippedKey) if (std::strncmp(key, skippedKey, sizeof(skippedKey)-1) == 0) return;
-  // OSM technical info tags
-  SKIP_KEY_BY_PREFIX("created_by");
-  SKIP_KEY_BY_PREFIX("source");
-  SKIP_KEY_BY_PREFIX("odbl");
-  SKIP_KEY_BY_PREFIX("note");
-  SKIP_KEY_BY_PREFIX("fixme");
-  SKIP_KEY_BY_PREFIX("iemv");
+std::string_view kUselessKeys[] = {
```
Comment (biodranik) [inline generator/osm_element.cpp:40]: constexpr or const?

File: generator/osm_element.cpp (hunk 1 above)
Comment (biodranik) [reply generator/osm_element.cpp:69]: There is a difference, of course. The reference version reads memory, the value version reads registers, and two fewer assembler lines: https://godbolt.org/z/351fvnhc4

The same for PointD, memory reads are replaced with register operations https://godbolt.org/z/KT9KPjzcq . So if we migrate to pass-by-value, 64 bit devices will definitely win. 32 bit ones at least work at the same speed:

выява

File: generator/osm_element.cpp (hunk 2 above)
Comment (biodranik) [reply generator/osm_element.cpp:74]: At least it is important for consistency. We don't want to have double standards, right? )

**Pull request merged**